### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by david1603**

### CODE REVIEW

File: hw04/Higher-Order.hs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+-- ex2
+type Height = Integer
+
+data Tree a = Leaf
+            | Node Height (Tree a) a (Tree a)
+    deriving (Show, Eq)
+
+foldTree :: [a] -> Tree a
+foldTree = foldr balancedAdd Leaf
```
Comment (Ivan1931) [inline hw04/Higher-Order.hs]: Very out of scope of this excersise but have a look at the [foldable type class](https://wiki.haskell.org/Typeclassopedia#Foldable)

File: hw04/Higher-Order.hs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+-- ex1
+fun1' :: [Integer] -> Integer
+fun1' = product . map (subtract 2) . filter even
+
+fun2' :: Integer -> Integer
+fun2' = sum . filter even . takeWhile (>1) .
+        iterate (\x -> if even x then div x 2 else 3 * x + 1)
+
+-- ex2
+type Height = Integer
```
Comment (Ivan1931) [inline hw04/Higher-Order.hs]: I feel like you could constrain this type even more - [numeric natural](http://hackage.haskell.org/package/base-4.8.1.0/docs/Numeric-Natural.html). While I don't think this is thaaaat necessary it does help using the types as documentation and could prevent all possible errors with a heigh being negative.

File: hw04/Higher-Order.hs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+data Tree a = Leaf
```
Comment (Ivan1931) [inline hw04/Higher-Order.hs]: You could make this datatype into an instance of functor which would help with some of the methods bellow :)

File: hw04/Higher-Order.hs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+-- calculate the height of a node in a binary tree given its two children
+newHeight :: Tree a -> Tree a -> Height
+newHeight left right = succ . foldr max 0 $ map getHeight [left, right]
+
+balancedAdd :: a -> Tree a -> Tree a
+balancedAdd x Leaf = Node 0 Leaf x Leaf
+balancedAdd x (Node 0 Leaf y Leaf) = Node 1 (Node 0 Leaf x Leaf) y Leaf
+balancedAdd x (Node 1 left_child y Leaf) = Node 1 left_child y (Node 0 Leaf x Leaf)
+balancedAdd x (Node _ left@(Node height_left _ _ _) y right@(Node height_right _ _ _))
```
Comment (Ivan1931) [inline hw04/Higher-Order.hs]: This really good haskell! Nice!

File: hw04/Higher-Order.hs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    | height_left < height_right = let new_left = balancedAdd x left
```
Comment (Ivan1931) [inline hw04/Higher-Order.hs]: I personally prefer to write let statements on their own lines - which is convention in elm. 
```haskell
let
   a = some statement
in
   another statement with a
```

File: hw04/Higher-Order.hs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        in Node (newHeight new_left right) new_left y right
+    | otherwise = let new_right = balancedAdd x right
+        in Node (newHeight left new_right) left y new_right
+
+-- ex3
+xor :: [Bool] -> Bool
+xor = foldr (\x y -> (x && not y) || (y && not x)) False
+
+map' :: (a -> b) -> [a] -> [b]
```
Comment (Ivan1931) [inline hw04/Higher-Order.hs]: You can avoid writing `map'` and write `map` instead by hiding the `Prelude` import of `map`. 
```haskell
import Prelude hiding (map)
```
will do that

File: hw04/Higher-Order.hs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+map' f = foldr (\x y -> f x : y) []
+
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f = foldr (flip f)
```
Comment (Ivan1931) [inline hw04/Higher-Order.hs]: This example breaks my brain even after all these years

File: hw04/Higher-Order.hs (hunk 2 above)
Comment (david1603) [inline hw04/Higher-Order.hs]: Didn't even know those were a thing, but definitely!

**Pull request merged**